### ISSUE TITLE
Req: Send requests

### ISSUE THREAD

**Issue by pklapperich** [therealklanni/slack-invite-request#4]
I was reading this blog: https://levels.io/slack-typeform-auto-invite-sign-ups/ which describes how to use the undocumented "users.admin.invite" as part of his explaination of this whole invite system he made.

What I think would be cool is to add support for SLACK_OUTWEBHOOK_TOKEN and SLACK_API_KEY. Users could then mention SIR in their invite channel and tell it to accept someone. Something like:

sir accept agent5
would make the appropriate https request to the e-mail address associated with agent5

sir accept address@example.com
would make the appropriate https request to invite "address@example.com" (some users share in their comment that they'd like to use a different e-mail then their google e-mail)

This would make it easier to accept invites on mobile and also easier to keep track of which invites have been sent out

**Comment by therealklanni** [comment]
That would definitely be cool, but it would probably have to be a "slash command", since SIR isn't a bot. Also, this would have to be configured separately from SIR, I think, since it's just a Slack feature and it would be initiated from Slack, so I don't see a need for the middle-man.

So, unless there's some reason we need SIR to act as a middle-man for that, at best we should include a how-to for setting it up in your Slack.

@pklapperich thoughts?

**Comment by pklapperich** [comment]
If SIR had an outgoing webhook, then it would receive messages sent to the channel, right? If there's a way to set up slack to send invites as slash command, I'd be supportive of that, too. The method described is just the first way I thought of to achieve the feature. I'm rather new to slack.

Taking a look at using slash commands, I have 2 concerns:
1. I think anyone in the group can issue any slash command, meaning a slash command to send invite e-mails might not be very secure. If SIR is listening to a private group, then only people in that group could interact with SIR, so one could control who can invite users based on who they add to the private group.
2. It looks like the data sent from slash commands is fixed, so something in the cloud (SIR? Some other webapp?) would still need to receive the GET/POST request generated by the slash command then then make the appropriate API request. I'm not sure this can be done entirely in slack itself.

I guess another reason having SIR act as a middle man might make sense is the "Planned" section of the README lists reddis and admin page. If SIR is initiating the invites, then that would be one good way for it to know who you invited. Or maybe this whole feature could just be postponed and added as part of the admin page.

**Comment by therealklanni** [comment]
Right, we **can** use SIR to handle that, as the middle man, but all SIR would do is send it right back to Slack. So to me it makes more sense to just configure the "slash command" to send it right to the Slack API instead.

You and/or I would have to look into whether that would be feasible of course. In theory it should work.

As far as the slash commands being available to anyone. I think you're right, they could. But we should look into it because it may be that the request is tied to the permissions of the user sending it, or we may be able to figure out a way to handle that. **Update**—I think if this is a problem, it will still be a problem even for posting from a webhook to SIR.

So, you may be right, maybe we _can't_ do it directly, as ideal as that would be, and in that case it would make sense to let SIR handle it. (Side note: I think I will want this functionality to be a plugin, rather than build into the core, since probably not everyone will want it).

Good point about the admin page, too. It's all definitely getting me brainstorming.

**Comment by pklapperich** [comment]
**Slash Command**
[Slash commands](https://api.slack.com/slash-commands) send the username and user ID as part of their post data. So SIR could then check the [user.info](https://api.slack.com/methods/users.info) method from the API to see if the person who issued the slash command is an admin. Finally, SIR could send the API request to invite the user.

_Pro:_ Makes more sense from an implementation standpoint. I mean, it's a command, right?
_Pro:_ Harder to invoke accidentally in conversation
_Con:_ Command and response from command are not visible to other users. But SIR could its existing inbound hook to post "Invite sent to foo@bar.baz"

**Outgoing Webhook**
Looking at [outgoing webhooks](https://api.slack.com/outgoing-webhooks), it looks like they don't support private groups (I thought they did). Outgoing webhooks can either a) listen to a specific global channel or b) listen to all channels and groups, but only messages containing keywords. The post data from an outgoing webhook contains the user ID and channel ID, so the same check outlined above for the webhook would need to be done. 

I'm not sure what the channel_id field shows if the message is sent from a private group. But in theory SIR could also check channel_id or channel_name and ignore messages not sent from the right group.

_Pro_ everyone in the channel sees the command. Interaction would look like talking to slackbot
_Pro-weak_ It seems to be pretty common for apps using outgoing integration also use incoming integration, so this route might be less likely to require an additional integration. (Since free accounts only get 5 by type)
_Con_ Setup slightly more involved (esp if SIR filters by channel_id; would require user figure out their channel_id and enter it into an ENV var).

**Slash Command sans SIR**
I don't think this is possible. We can't control the post data sent by a slash command and that post data isn't what the API request requires. Something needs to catch the slash command post message, parse the text field, and construct and post the API request for invite.

Regarding your side note... are you thinking of adding a plugin API to SIR? Or are you more envisioning a freestanding herokuapp that's associated with SIR and referenced by SIR's readme?

**Comment by therealklanni** [comment]
Excellent job researching this. Seems like Slash commands via SIR are going to be the best option after all. Thanks for the hard work looking into this.

As far as the con (response from command are not visible to other users), posting a new message to the same channel that invite requests appear in, saying so-n-so was invited by whoever seems like a good solution to me. "sonso@somewhere.net was invited by whoever"

This will be a fairly big implementation and I think I should address the non-persistent uploads bug first, so it might be a while before this feature is completed. That is, unless anyone else can contribute to help get this done. :metal:
